### ISSUE TITLE
Build tool to import firebase JSON backup to MySQL

### ISSUE THREAD

**Issue by snickell** [code-dot-org/code-dot-org#55189]
Firebase contains a terabyte of data. The easiest way to slurp it we've found is using one of the firebase backups, which is a giant terabytes JSON file. This presents challenges, since we need to streaming parse the JSON since it won't fit in RAM.

See: https://github.com/code-dot-org/firebase-migration-tool

**Comment by snickell** [comment]
After trying and failing to parse this scale of data with the first-pick ruby streaming parser gem, we're trying https://github.com/uhop/stream-json which seems popular in the nodejs world and is supposed to have good performance.

See: https://github.com/code-dot-org/code-dot-org/tree/unfirebase/unfirebase

We've discovered a few caveats:
1. We're only getting about 10MB/s speed on (streaming parse json => mysql insert), with the bottleneck appearing to be the streaming JSON parse not the mysql insert.
2. This means we're looking at ~24hrs+ for a full import
3. Projects can currently use an arbitrary (? check firebase rules) amount of disk space. As a result, we're having to run `'SET GLOBAL max_allowed_packet=10000000000;'` to support 8MB+ inserts. Unfortunately this sets max_allowed_packet for the lifetime of the mysql instance. We couldn't find a way around this.

**Comment by snickell** [comment]
If we want a faster parse, we may need to consider a parser written in a lower level language like C/C++/Rust. Unfortunately we require not just speed, but a streaming API since we can't hold a terabyte in RAM. 

A general overview of really fast JSON parser perf: https://github.com/simdjson/simdjson#performance-results suggests anywhere from 0.5GB/s (50x current) to 7GB/s is achievable, with the major caveat that these charts are for the NON streaming form of the API.

If we're willing to use C++, we could use the "event based parser" approach to the simdjson API: https://simdjson.github.io/simdjson/md_doc_ondemand_design.html . Or the simpler (and possibly slower, but at some point DB perf is gonna be the bottleneck) SAX API from RapidJSON: https://rapidjson.org/md_doc_sax.html

With our existing `stream-json` based parser, we get **13MB/s without SQL** inserts (=153s on the 2GB big-data.json) and **10MB/s with SQL inserts** (=205s on the 2GB big-data.json) .

**Comment by snickell** [comment]
Implemented a C++ "sax style" parser using RapidJSON (https://github.com/code-dot-org/code-dot-org/blob/unfirebase/unfirebase/parse-firebase-json.cpp) and we **get 285MB/s without SQL** (=7s on the 2GB big-data.json).

This seems like a no-brainer assuming the corresponding MySQL code in C++ isn't too big a pain, and we can get SQL perf high enough that its not the primary bottleneck.

**Comment by snickell** [comment]
Learning the hard way: we can't use the modern/documented C++ API to MySQL because it requires the "X plugin" to be enabled
1. This plugin is explicitly not supported by AWS Aurora
2. Its not on by default for MySQL 5.7 anyway

As a result, we're using the older/crunchier MySQL "JDBC" api, which requires some homebrew tweaking to even work.

Doing a parse + synthetic inserts (not real data, and probably much smaller strings, but same number of inserts) we get **240MB/s with synthetic SQL** (=8.3s on the 2GB big-data.json).

**Comment by snickell** [comment]
With the parser collecting the real value strings, we see a significant drop compared to "285MB/s without SQL" to **142MB/s extracting the string form of each record without SQL** (=14s on the 2GB big-data.json) using a dynamically switched rapidjson::Writer to collect the string, but only doing the small/synthetic insert. This could probably be optimized.

Adding SQL inserts of the real data onto that, we see: **65MB/s for real full SQL values to unfirebase schema** (=32s on the 2GB big-data.json).

So that's our realest value for C so far. For the whole 1 terabyte file, if we don't optimize the C further, we can expect 4-5hrs for a firebase import (to local MySQL, we still have to check RDS Aurora from EC2).

Right now, with the C++ version, the bottleneck is very much the large (up to 4MB) "INSERT INTO" statements. We've already done most of the optimizations from https://dev.mysql.com/doc/refman/5.7/en/insert-optimization.html, but allegedly using `LOAD DATA LOCAL INFILE` is MUCH faster (multiples faster?) than insert statements (!!!), and can transmit local files to the remote server, so it should work with RDS.

Ideally we could get our speed with MySQL in the loop up to the 240MB/s parse, which would finish the 1 terabyte in about an hour.

**Comment by snickell** [comment]
OK! I'm now using [a] background thread(s) to do the bulk `LOAD DATA LOCAL INFILE` calls. We build them up the data to be inserted as TSV file chunks, and queue them for the background thread to do the network IO copying to the MySQL daemon, and then waiting for it to parse and import.

With one background thread, on my machine (benchmark for the values above), with real data going to local mysqld, we see: **125MB/s with SQL** (=16s for 2GB big-data.json)

I've also gotten this compiling on Linux now, and **running on EC2 against RDS I'm getting half that performance, I get 61 MB/s** (=34s for a 2GB big-data.json, though not the /exact/ same one)

It looks like:
1. RDS filesystem is slower (maybe use a local NVMe SSD instead of EBS volume?)
2. The network IO is a major block.... possibly using two background threads would help?
3. Maybe the RDS instance can be made beefier/faster?

**Comment by snickell** [comment]
Using 4x background threads **running on EC2 against RDS I get 98MB/s** (=21s for big-data.json). Using 8x (21s) was the same speed, and 4x (21s) was a tiny bit faster than 2x (23s), which are all way faster than 1x (34s).... really though you just need 2x in this case to get 75% of the gains.

**Comment by snickell** [comment]
It looks like we're bandwidth constrained at the moment (98MB/s consistently is suspiciously close to a 1gigabit limit, possibly on EBS in bandwidth, possibly on outgoing bandwidth to RDS). So....

**LETS GO SHOPPING!!!**

Here's the three tiers that made sense to us based on price, bandwidth, and EBS bandwidth / ssd presence (thanks: https://instances.vantage.sh/):

1. If we want to go all-in on Bandwidth, and not use NVMe: `c7gn.8xlarge, 16667.0 Mbps EBS, 100 Gigabit, $2/hr`
2. Big step up with huge NVMe capability and reserved 10Gigabit ethernet: `c5ad.8xlarge, 1200 GB (2 * 600 GB NVMe SSD) + 3170 Mbps EBS, 10 Gigabit, , $1.38/hr`
3. Basic model, what we should use for dev, might be fast enough for the real deal?: `m5.2xlarge, 2300.0 Mbps EBS, "up to 10 Gigabit", $0.38/hr`
4. Current adhoc instance for **above benchmarks-to-RDS speeds**: `t2.2xlarge, EBS only but NO EBS BW value, "Moderate" network perf (apparently means 300-900 Mbps in AWS-speak", $0.37 hour

Question: should we switch the adhoc instance to m5.2xlarge??? It seems better all around, same RAM, reserved 8vCPUs instead of "burst up to v8".

**Comment by snickell** [comment]
We tried m5.2xlarge but got the same perf. We then realized we were using a "general purpose" gp2 EBS storage type. These max out at at very low EBS bandwidth: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volume-types.html

Using an m5 instance we can use "nitro" storage, so we're switching to the fastest device type at present: io2 block express


Here's the settings we're modifying to:
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/90983ba9-e6de-4dec-8536-a61749e55e8f)



If this doesn't work, we'll try using a local NVMe (much more $).

**Comment by snickell** [comment]
Temporary setback, we need to increase our io2 limit in AWS:
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/19e5c3f8-bbd1-445a-b223-82d643bfabc7)

**Comment by snickell** [comment]
gp2, our default adhoc EBS volume type: `General Purpose SSD (gp2) Volumes | $0.10 per GB-month of provisioned storage`
= $0.1 * 1000GB = **$100/month for gp2**

io2 EBS volumes: `$0.032/provisioned IOPS-month for greater than 64,000 IOPS`
= $0.032 * 100,000 IOPS (up to 256,000) = **$3.2k/month for io2** (up to $8k/month)

So.... that's crazy bling expensive, maybe we DON'T want to use this kind of volume 😰, but 10k IOPs is $300/mo, not a huge huge deal, and I don't think we're using ANNNNYWHERE like 10k IOPs.

For comparison, here's our (prospective new prod sized, with 18k IOPs guaranteed) **unfirebase RDS cluster's IOPS metrics** during one of our mass-import attempts at 100MB/s from EC2. We see spikes of 1.2k IOPS during hard importing, so about 7% utilization.... 10k IOPS is a LOT in our workload, the read side is going to be even lower number per second!:
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/93627956-9a3e-4006-94ee-7f6604a1746b)

**Comment by snickell** [comment]
Maybe our bottleneck is outgoing IO bandwidth? Here's metrics from our EC2 instance at both sizes... its claiming we're moving 6GB/s out on the network card... I dunno how that's possible since we're transmitting 100MB/s uncompressed JSON to RDS using a supposedly very efficient transfer method???
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/ca581fb1-24a6-49cf-acf5-d63b7d02608a)

Firstly, these instances are supposed to cap at "up to 10GBit" = 1GB/s, so how in the world is this happening? I'm suspicious of this graph, but it might be one to dig into, especially if changing the EBS volume type shows no improvement.

**Comment by snickell** [comment]
big-data.json resulst are in for EC2 with an io2 volume to RDS, and, curiously, switching to the io2 volume type has **not shown a huge improvement**! In short **using a faster EC2 instance with a MUCH faster EBS volume had NO EFFECT**, at least not on big-data.json runs.

**100MB/s for EC2 m5.2xlarge with an io2 volume to RDS** (="21s using 2GB big-data.json").

Notice this is IDENTICAL to our earlier EC2 values using the slower/default-adhoc instance type, and a much slower gp2 volume to RDS: https://github.com/code-dot-org/code-dot-org/issues/55189#issuecomment-1830659981
> EC2 against RDS I get 98MB/s (=21s for big-data.json)

Again, consistent 98MB/s is suspiciously identical to 100MB/s. I'm not sure I trust the network-out bytes graph in EC2....

**Comment by snickell** [comment]
Bottleneck is not EBS volume instance IOPs. Write OPS are neglible, and here's a graph of read IOPs with (slow instance with slow gp2 ebs) vs (fast instance with ultra-fast io2 ebse). Notice they're basically the same, it looks like **EBS Volume IOPs are not the bottleneck**.

Also, since I'm at it, for future reference we probably **only need a couple thousand IOPs to run this job**:
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/71e567cd-5a65-433e-af0f-6d2b104fbfb7)

**Comment by snickell** [comment]
We see a VERY interesting story on the EBS volume read throughput (KB/s) monitoring graph:
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/7a2b7033-c4ba-44c2-9966-3c43c8570282)

This graph shows the same 100MB/s limit we're measuring in our 98MB/s big-data 2GB test!!!

Is this the smoking gun? Notice that **we see this 100MB/s cap with both the gp2 "slow ebs volume type" (volume type supposed to be rated for 250 MiB/s, instance type unreated) AND with the io2 "ultra-fast ebs volume type" (volume type supposed to be rated for 4GB/s, instance type rated for min 250MB/s to max500MBs)**.

1. This suggests that the cap is not EBS volume throughput, even though this is the first graph I've found that shows the same limit we experience.
2. So what is the limit?

**Comment by snickell** [comment]
## Benchmarks of EBS vs local NVMe SSDs

**EBS** c5ad gp2
  - **READ: bw=250MiB/s** (262MB/s), 250MiB/s-250MiB/s (262MB/s-262MB/s), io=10.0GiB (10.7GB), run=40974-40974msec
  - **WRITE: bw=257MiB/s** (269MB/s), 257MiB/s-257MiB/s (269MB/s-269MB/s), io=10.0GiB (10.7GB), run=39902-39902msec

**SSD** c5ad (2x nvme llvm data pool)
  - **READ: bw=660MiB/s** (692MB/s), 660MiB/s-660MiB/s (692MB/s-692MB/s), io=10.0GiB (10.7GB), run=15526-15526msec
  - **WRITE: bw=172MiB/s** (181MB/s), 172MiB/s-172MiB/s (181MB/s-181MB/s), io=10.0GiB (10.7GB), run=59460-59460msec

```
# benchmark bulk read   
fio --name TEST --eta-newline=5s --filename=fio-tempfile.dat --rw=read --size=500m --io_size=10g --blocksize=1024k --ioengine=libaio --fsync=10000 --iodepth=32 --direct=1 --numjobs=1 --runtime=60 --group_reporting

# benchmark bulkwrite
fio --name TEST --eta-newline=5s --filename=fio-tempfile.dat --rw=write --size=500m --io_size=10g --blocksize=1024k --ioengine=libaio --fsync=10000 --iodepth=32 --direct=1 --numjobs=1 --runtime=60 --group_reporting
```

## Just getting the firebase dump is SLOW

- Downloading the firebase.json.gz file from Google Storage: 6.8 minutes
- GUnzipping the file: 77 minutes (!!!)

**Comment by snickell** [comment]
...And we still have an overflow bug... left the adhoc running while I slept and came back to a quit at 15% with a suspiciously overflow-y looking last few tsvs: [997000, 998000, 999000.... ICRASHQUIT]:

```
Written to MySQL: 155.12GB, ~15.12%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/992000.tsv'
Written to MySQL: 155.27GB, ~15.13%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/993000.tsv'
Written to MySQL: 155.42GB, ~15.15%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/994000.tsv'
Written to MySQL: 155.6GB, ~15.17%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/995000.tsv'
Written to MySQL: 155.75GB, ~15.18%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/996000.tsv'
Written to MySQL: 155.92GB, ~15.2%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/997000.tsv'
Written to MySQL: 156.08GB, ~15.21%
Written to MySQL: 156.25GB, ~15.23%
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/998000.tsv'
LOAD DATA LOCAL INFILE '/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/999000.tsv'
Written to MySQL: 156.31GB, ~15.23%
loadDataThread: done=true, draining queue
loadDataThread: done
Written to MySQL: 156.46GB, ~15.25%
loadDataThread: done=true, draining queue
loadDataThread: done
Written to MySQL: 156.63GB, ~15.27%
loadDataThread: done=true, draining queue
loadDataThread: done
Written to MySQL: 156.8GB, ~15.28%
loadDataThread: done=true, draining queue
loadDataThread: done
DONE with parse-firebase-json!
ubuntu@adhoc-unfirebase:~/adhoc/unfirebase$ 
```

Crashing on a multiple of 10 means that its a string related bug, whereas anything binary would have overflowed at a power of 2. My guess is we have some crazy C++ strings copying pointers when they should be duplicated.... have to learn more C++ memory model 🚱 

One final clue and back to 💤 :
```
>>> len('/scratch/tsvs//scratch/prod-firebase-data.json-tsvs/999000.tsv') + 1 # one extra byte because C ter
minuates strings with a NULL
63
>>> # 63! That means the next character will be a zero-indexed 64 and break through a power of 2
```

I'm gonna restart it with a shorter / less screwy filename. Might just work to 100%.

**Comment by cnbrenci** [comment]
Today, trying to track down overflow bug

Ran with higher numbers in the filename which worked fine. Seems to be something with the writer in the parser. If one of the parsing fxns returns false, it fails silently. 

Currently doing a run that should print out the last keys encountered if we run into a scenario where the writer is returning true from its method calls. Hopefully we can get some more clues from the result of the run, but it will take some time to hit the failure.

**Comment by snickell** [comment]
Based on the following 4 test runs, it looks like neither threads nor writers nor SQL are involved (cuts a lot of code to analyze!):
1. Just checked the result of the 'last keys' test run: to my surprise NONE of the `writer->Whatever()` calls are returning `false` (and it still bailed in the same place), so back to the drawing board.
2. As a followup tried a test run with `USE_WRITER_TO_COLLECT_STRING=false`, which disables writers entirely.
    - It still failed in the same way at `999000.tsv`, so its definitely not writers, and its not related to the amount of data transmitted to the RDB or the size of the TSVs (both are much smaller without writers, it uses tiny fake strings instead).
    - On the upside, the runs are much faster with writers disabled, so we at least have a repro thats 3x or whatever faster.
3. Also tried a run with `LOAD_DATA_IN_THREAD=false` and `USE_WRITER_TO_COLLECT_STRING=false`, it failed too.
4. Also tried disabling the SQL `LOAD DATA` call, still failed.

As another experiment, I tried setting `NUMBER_OF_RECORDS_BEFORE_COMMIT = 1100;` (vs 1000) to get a different cadence / pattern to the filenames with the same data. So rather than each filename being prev+1000 its now prev+1100. The result was it stopped on almost the same "channel" (a record is one "channel", i.e one firebase project, in this case).... BUT! It stopped on 998800, NOT 999900 (=998800 + 1100). So its not an overflow on the filename.
```
LOAD DATA LOCAL INFILE '/scratch/prod-firebase-data.json-tsvs/995500.tsv'
Written to MySQL: 0.07GB, ~0.01%
LOAD DATA LOCAL INFILE '/scratch/prod-firebase-data.json-tsvs/996600.tsv'
Written to MySQL: 0.07GB, ~0.01%
LOAD DATA LOCAL INFILE '/scratch/prod-firebase-data.json-tsvs/997700.tsv'
Written to MySQL: 0.07GB, ~0.01%
LOAD DATA LOCAL INFILE '/scratch/prod-firebase-data.json-tsvs/998800.tsv'
Written to MySQL: 0.07GB, ~0.01%
DONE with parse-firebase-json!
```

**Comment by snickell** [comment]
On the performance bottleneck, just want to document that we see **130MB/s with an EC2 m5ad parsing from the super fast SSD**. So that's a little faster with the SSD, but not much.

We miiight be CPU bound, see `top -H` the main thread is 100% pegged, but I think that might be an artifact of the lockless queue structure I'm using (it keeps polling = pegs CPU):
![Image](https://github.com/code-dot-org/code-dot-org/assets/223277/c35df24f-1f80-49e0-8d38-1aea52d243be)

Hint that we're NOT CPU bound is that `Num data jobs queued: 11`, i.e. background threads still aren't catching up to the main thread....

So at this point (reading from 6x faster local SSD, only getting a little more bandwidth), I think **the bottleneck must in RDS, or our networking, or the network path between us and RDS**.

**Comment by snickell** [comment]
> So its not an overflow on the filename.

Following up on this, I printed the actual last record number being parsed:
```
Total channels parsed: 999351
Last channelID: 8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE
Were we in the middle of a channel when we quit? true
```
So there's something in the **middle** of the parse for channel 999351 (id=8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE) that causes it to bail...

This looks like a likely candidate for a bad/invalid-in-json character breaking rapidjson (uhoh, we might need to patch it?!?):
1. `8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE` is a chat app
2. Lots of upper unicode characters are being used prior to the parser terminating.
![image](https://github.com/code-dot-org/code-dot-org/assets/223277/6b863858-50e0-4c9c-b14e-9951e7546289)

**Comment by snickell** [comment]
This looks promising, RapidJSON explicitly bails if it hits the NUL character (`\0`): https://github.com/Tencent/rapidjson/blob/0e8bbe5e3ef375e7f052f556878be0bd79e9062d/include/rapidjson/reader.h#L713

Maybe firebase doesn't parse them out?

**Comment by snickell** [comment]
TL;DR: the parse appears to have failed on a JSON key name that has 2 sequences of 3 bytes (0xEF 0xB8 0x8F) that are **not valid UTF-8**.

I found the offset for this section of the huge file, and extracted the next bytes after the parse failed. 

The highlighted bit is the last sequence that parsed, and the next key is 🥁:
<img width="555" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/d8083f9d-23d1-4b19-a0fb-93860e993199">

I believe this is the offending sequence (its literally the next key after parse dies):
```
"takenusernamepiper 👁️\U0001fae6👁️":"true",
```

Viewed in Hex Fiend (a hex editor), the hex sequence of chars is:
<img width="455" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/aec0834f-78fe-48a8-bf65-0af896de5a67">

Looking at the non-ASCII characters (dots in the hex fiend text view):

UTF Char 1 takes 4 bytes and its an 👁️ emoji: 
<img width="454" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/65f87509-ae50-4bcf-901d-ad4b075812b5">

The next 3 byte sequence, **0xEF 0xB8 0x8F is NOT VALID UTF-8**:
<img width="456" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/6e1b3961-41b9-4b97-85cc-06831d9f9425">

I'm skipping `\U001fae6` as this is all boring ascii: its not actually a `\U001` character in the key, that's actually 5 valid UTF-8 characters, i.e. `['\', 'U', '0', '0', '1']`.

The next 4 byte sequence is another 👁️ emoji:
<img width="456" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/502728ce-0a68-4c36-9cba-abe79b3fe989">

And finally we have another **0xEF 0xB8 0x8F is NOT VALID UTF-8**
<img width="458" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/6c85c9de-f0ba-4f4c-a8d4-969e62db5946">

Notice that in the screenshot, the offending 3 bytes sequence doen't show up at ALL, its just not printable:
<img width="112" alt="image" src="https://github.com/code-dot-org/code-dot-org/assets/223277/032d3ef4-f311-4c35-a975-ebcfbc4c7535">

**Comment by snickell** [comment]
So `EFB88F` is apparently at UTF-8-ization of the "variation selector", which can be used to tell the computer to render a pre-emoji character as an emoji: https://twitter.com/RobbyKraft/status/971534867426144257?lang=en

https://www.fileformat.info/info/unicode/char/fe0f/index.htm

In theory this might be valid UTF-8 but very odd.... in any case the hex editor didn't like it so who knows. Time to try it and see what happens to rapidjson!

**Comment by snickell** [comment]
Nope! I had it wrong:
1) UTF-8 escape sequences _are_ part of JSON strings, and are interpreted.
2) `\u0001` _is_ a valid UTF-8 escape sequence, and RapidJSON allows it in strings.
3) HOWEVER, `\U0001` is **not valid** (uppercase U), and adding ANY `\U####` sequence to any JSON causes RapidJSON to flip shit on it.

So that's the problem, we got [WUT]ed with some invalid UTF-8 escape sequence that Firebase let through for whatevs reason, but RapidJSON no likes.

**Comment by snickell** [comment]
It appears when using the `c5ad` instances we may be CPU bottlenecked. Its a _little_ hard to tell because the behavior of the boost lockfree queue has the producer spin its CPU when blocked.

We're looking for an instance:
- With high single thread perf (higher than a 3.3GHz epyc)
- With a`fio` benchmark showing read off EBS beyond 250MB/s
- With network performance > "10 Gigabit" (in reality you don't get a single connection that saturates this with the exception of two instances in the same placement group, but higher gives you more according to an undisclosed formula)

Best match might be `m5zn.6xlarge` (https://instances.vantage.sh/aws/ec2/m5zn.6xlarge):
- Designed explicitly for single-thread perf (4.5GHz xeon platinum 8252)
- 50 gigabit networking
- 1GB/s baseline throughput to EBS
- $2/hr

Matching EBS volume, using gp3:
- 2TB
- 6000 IOPs provisioned
- 1GB/s provisioned bandwidth
- $210/mo (using monthly because still billed while off)

**Comment by snickell** [comment]
Without pushing to MySQL instance, we now get 150MB/s using m5zn.6xlarge and gp3.

With pushing to MySQL we get a very disappointing 70MB/s.

**Comment by snickell** [comment]
`parse-firebase-json.cpp` has been extended to give the byte offset where the error occurred, and to dump the IO context before and after this offset to disk:
```
JSON parse error at byte offset 156137170691, code 10: Invalid escape character in string.

Were we in the middle of a channel when we quit? true
Wrote parsed channel JSON so far to: error-channel-write-buffer-8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE.json
lastKeys: .v3.channels.8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE.storage.keys.takenusernamepawlohan000.
Last channelID: 8iMQlhkrfx_SwHNtvBDwm2fXx3EyBhirTLH49vKKBcE
Current Record Number: 996429
Wrote file contents before error byte offset to: error-bytes-before.json
Wrote file contents after error byte offset to: error-bytes-after.json
```

**Comment by snickell** [comment]
The RapidJSON code throwing the "inavlid escape character" error is from `GenericReader::ParseStringToStream`: https://github.com/Tencent/rapidjson/blob/f9d53419e912910fd8fa57d5705fa41425428c35/include/rapidjson/reader.h#L1047

The `Reader` type we declare is just `typedef GenericReader<UTF8<>, UTF8<> > Reader` (see: https://github.com/Tencent/rapidjson/blob/f9d53419e912910fd8fa57d5705fa41425428c35/include/rapidjson/reader.h#L2233).

So we could potentially create a `IgnoreEscapeReader` subclass of GenericReader and override ParseStringToStream to skip (or munge) certain invalid escape codes.

**Comment by cnbrenci** [comment]
We are trying to test out some different schemas, running into an error importing the entire channel into a json column, error `Cannot create a JSON value from a string with CHARACTER SET 'binary'. (MySQL error code: 3144, SQLState: 22032 )`

Not finding a good solution so we're pivoting to trying out a schema with one record per row and a composite key including (ChannelId, TableName, RecordId). 

Next steps are going to be:
- Define schema for these 3 composite keys and a text field for the row
- Update parse-firebase-json.cpp path with LOAD_DATA_INSTEAD_OF_INSERT = false
- Keep track of the bits of info for the composite key and insert them along with each record in the records list.

Notes:
- Careful not to overwrite the existing DB
- adhoc version of script might have different changes at the moment than branch
- optional - try json field for the row instead of text and see if we run into the same issues as before?

**Comment by snickell** [comment]
Current Firebase limits:
```
[production] dashboard > CDO.firebase_max_record_size
=> 4096
[production] dashboard > CDO.firebase_max_property_size
=> 4096
[production] dashboard > CDO.firebase_max_table_rows
=> 20000
[production] dashboard > CDO.firebase_max_table_count
=> 10
[production] dashboard > CDO.firebase_max_channel_writes_per_15_sec
=> 300
[production] dashboard > CDO.firebase_max_channel_writes_per_60_sec
=> 600
[production] dashboard > CDO.firebase_max_record_size * CDO.firebase_max_table_rows
=> 81920000
```

So up to 81MB in a table, and up to 800MB for a channel (10x tables)

**Comment by cnbrenci** [comment]
Current open questions: 
What are the potential gains from deduplicating data (because many tables are stock tables)
What are the performance trade offs between the three modes of data storage?

**Comment by snickell** [comment]
Obviously we're gonna have to revisit this tool as the last step when we actually import, but for now we've loaded 100% of the data in the "channel JSON blob" format, and we'll be using the data imported by the tool to determine a good schema shape going forward.

**Comment by snickell** [comment]
> So we could potentially create a IgnoreEscapeReader subclass of GenericReader and override ParseStringToStream to skip (or munge) certain invalid escape codes.

We ended up taking this approach, see `munged-reader.h`
